Added ability to store results from theano function in a python dictiona...

DeathMonster666 opened the pull request:
...ry

DeathMonster666 commented:
Does not support named input arguments.

abergeron commented:
We don't use CamelCase for files and function names.  Also this function looks simple enough to be added directly to `theano/compile/function.py` no need for a new file.

Finally make sure to run flake8 and fix any errors it gives for new code.

DeathMonster666 commented:
I removed CamelCase and moved the code back into function.py.  I also ran flake8, and fixed the style issues (which were all related to extra lines).  

I also changed the code so that the output list is sorted alphabetically based on the keys provided by the user.

nouiz commented on theano/compile/function.py:
```diff
@@ -185,6 +195,24 @@ def opt_log1p(node):
+            output_keys.append(pair[0])
+            outputs.append(pair[1])
+
+
```
only one empty line.

nouiz commented on theano/compile/function.py:
```diff
@@ -185,6 +195,24 @@ def opt_log1p(node):
+        outputs_dict_format = True
+        output_items = outputs.items()
+
+        output_items_sorted = sorted(output_items, key = lambda x: x[0])
```
no space around argument keywork (that mean no space around the = of key)

nouiz commented on theano/compile/function.py:
```diff
@@ -267,4 +295,16 @@ def opt_log1p(node):
+
+    if outputs_dict_format:
+
+        fn_dict_output = (lambda *args, **kwargs: output_dictionary_wrapper(args, kwargs, fn = fn, keys = output_keys))
```
I think this work, but I think it break the pickling of Theano function that use this feature. Also, it break the assumption that what we return is an instance of Function and so it do not have the attribute maker for example.

I think this code should be done in the method Function.__call__ directly. That should fix the pickling problem and keep the same object with the same attribute.

Can you add a test that make sure we do not break this mean feature?

thanks

DeathMonster666 replied:
Thanks.  I see where I can put it into Function.call, but adding it will require additional arguments for the Function class.  

So the current header: 

def __init__(self, fn, input_storage, output_storage, indices, outputs, defaults, unpack_single, return_none, maker):

would become: 

def __init__(self, fn, input_storage, output_storage, indices, outputs, defaults, unpack_single, return_none, maker,output_to_dictionary,output_keys):

I'll code and commit this change just to see that it works, but I imagine that adding too much configuration to the constructor arguments could hurt readability.

DeathMonster666 commented:
Updates to fix issues associated with Nouiz's comments: 

-Moved functionality into Function.call
-Added simple unit tests

DeathMonster666 commented:
One thing that I don't know how to test is non-determinism across different interpreter sessions in the "sorted" method.  

To be on the safe side, I'd be willing to throw an exception if the keys aren't strings or integers.

delallea commented on theano/compile/function.py:
```diff
@@ -185,6 +185,24 @@ def opt_log1p(node):
 
 
     """
+    if type(outputs) is dict:
```
Better use ``isinstance(outputs, dict)`` to work with subclasses as well.

delallea commented on theano/compile/function.py:
```diff
@@ -185,6 +185,24 @@ def opt_log1p(node):
+        outputs_dict_format = True
+        output_items = outputs.items()
+
+        output_items_sorted = sorted(output_items, key=lambda x: x[0])
```
You don't need to pass ``key`` here, tuples are sorted in lexicographic order.

delallea commented on theano/compile/function.py:
```diff
@@ -263,7 +281,9 @@ def opt_log1p(node):
-                profile=profile)
+                profile=profile,
+                output_dictionary_flag=outputs_dict_format,
+                output_keys=output_keys)
```
Couldn't you just pass ``output_keys`` to avoid having 2 arguments?

delallea commented on theano/compile/function_module.py:
```diff
@@ -280,7 +280,8 @@ class Function(object):
     def __init__(self, fn, input_storage, output_storage, indices, outputs,
-                 defaults, unpack_single, return_none, maker):
+                 defaults, unpack_single, return_none, output_dictionary_flag, 
+                 output_keys, maker):
```
Might be safe here, but in general try adding new arguments as last arguments, to be sure not to break any existing code.

delallea commented on theano/compile/function_module.py:
```diff
@@ -676,6 +679,17 @@ def __call__(self, *args, **kwargs):
+                for i in range(0, len(self.output_keys)): 
+                    outputDict[self.output_keys[i]] = outputs[i]
+
+                outputs = outputDict
```
Looks to me like you're re-building the dictionary you had in the first place... Why not just provide this dictionary in ``outputs``? (NB: I  haven't looked closely enough at the code to verify it would work)

DeathMonster666 replied:
"Looks to me like you're re-building the dictionary you had in the first place... Why not just provide this dictionary in outputs? (NB: I haven't looked closely enough at the code to verify it would work)"

The "outputs" passed into the dictionary is a map from the keys to theano expressions for the output variables.  Where I construct the outputs dictionary in Function.call, the values in the dictionary are the actual values that will be returned to the user.

delallea replied:
I was just wondering if maybe you could have simplified things a bit by just storing the dict in ``outputs`` instead if adding a new argument. But I had a quick look and I'm not sure it would actually be simpler since there's a bunch of code relying on the fact ``outputs`` is a list. So forget about it :)

Just two small code style points to fix:
1. Use ``is not None`` instead of ``!= None``
2. Your ``for`` loop can be written as: ``outputDict = dict(itertools.izip(self.output_keys, outputs))`` (actually you can just do ``return dict(...)`` without going through intermediary variables ``outputDict`` and ``outputs``)

DeathMonster666 commented:
Hello, 

I fixed the issues that have been brought up, but there are still a few issues with the code: 

-Doesn't work with debug mode.  This is why Travis CI is failing.  
-It looks like sorting with strings and integers works in python 2 but breaks in python 3.  

Looking into fixing things with debug mode.

nouiz commented on theano/tests/test_dictionary_output.py:
```diff
@@ -0,0 +1,117 @@
+        assert outputs['1'] == 40.0
+        assert outputs['c'] == 20.0
+
+    def test_input_dictionary(self):
```
change method name to reflect what it do. like test_input_kwargs

nouiz commented on theano/compile/function_module.py:
```diff
@@ -676,6 +676,18 @@ def __call__(self, *args, **kwargs):
+                assert len(self.output_keys) == len(outputs)
+
+                #for i in range(0, len(self.output_keys)): 
+                #    outputDict[self.output_keys[i]] = outputs[i]
```
remove this deadcode.

nouiz commented on theano/tests/test_dictionary_output.py:
```diff
@@ -0,0 +1,117 @@
+        assert result[0] == 47.0
+        assert result[1] == 420.0
+
+    def test_output_list_still_works(self):
```
this seem a duplicate test of test_output_list

nouiz commented on theano/compile/function.py:
```diff
@@ -185,6 +185,22 @@ def opt_log1p(node):
+    if isinstance(outputs, dict):
+        output_items = outputs.items()
+
+        output_items_sorted = sorted(output_items)
```
This still work on non string. Personally, I'm ok to let this, but @lamblin didn't liked it. Can you assert that all key are int or string? Was there a consensus that I missed on how to handle this? Otherwise the PR seem ok except my small comments. thanks

lamblin replied:
My worry is that the behaviour of `sorted` is not deterministic between runs.
I think we should restrict it to strings, at least for python 2 (maybe python 3 will raise an exception if the objects are not directly comparable).

DeathMonster666 replied:
In the latest version of the code I restricted it to strings.  Python3 does indeed raise an exception when sorting a list containing both strings and integers.  

The only non-string type that I could see some value in using for keys is a tuple.  For example: 

{("h", "dropout") : h_dropout, ("h", "no dropout"): h_no_dropout}

but I don't think that this is a big deal.

nouiz commented on theano/compile/function_module.py:
```diff
@@ -676,6 +676,18 @@ def __call__(self, *args, **kwargs):
+                #for i in range(0, len(self.output_keys)): 
+                #    outputDict[self.output_keys[i]] = outputs[i]
+
+                return dict(itertools.izip(self.output_keys, outputs))
```
Do you know which version is faster? the dead code or this version? Personnaly, I would prefer the faster, as otherwise it would add an extra overhead.

DeathMonster666 replied:
They're equally fast, so I went with the shorter one.  In either case, it's unlikely to be a bottleneck in function calls unless one is using a very large number of outputs, which is a bad idea anyway.

nouiz replied:
ok. it is great that you timmed it. You can be surprised of where overhead come from. For small Theano function like on scalar, the parsing of the input is a significant overhead...

delallea commented on theano/compile/function.py:
```diff
@@ -185,6 +185,22 @@ def opt_log1p(node):
+        output_keys = []
+        outputs = []
+        for pair in output_items_sorted: 
+            assert isinstance(pair[0], str)
```
Use ``basestring`` instead of ``str`` (unicode is ok).

lamblin commented on theano/compile/debugmode.py:
```diff
@@ -2179,7 +2179,8 @@ def __init__(self, inputs, outputs, optimizer, mode,
             profile=None,
-            on_unused_input=None):
+            on_unused_input=None,
+            output_keys=None):
```
Can you add documentation for "output_keys" in the docstring?

lamblin commented on theano/compile/function.py:
```diff
@@ -185,6 +185,22 @@ def opt_log1p(node):
 
 
     """
+    if isinstance(outputs, dict):
```
Update the docstring to tell that outputs can be a dictionary, and what the requirements (wrt sorting, etc.) are.

lamblin commented on theano/compile/function_module.py:
```diff
@@ -676,6 +676,18 @@ def __call__(self, *args, **kwargs):
         else:
+
+            if self.output_keys is not None: 
+                outputDict = {}
```
Use an OrderedDict here.

lamblin commented on theano/compile/function_module.py:
```diff
@@ -1396,7 +1410,8 @@ def register_checker(checker):
 def orig_function(inputs, outputs, mode=None, accept_inplace=False,
-                  name=None, profile=None, on_unused_input=None):
+                  name=None, profile=None, on_unused_input=None,
+                  output_keys=None):
```
Add documentation for `output_keys`.

lamblin commented on theano/compile/pfunc.py:
```diff
@@ -336,7 +336,7 @@ def __init__(self, variable, default=None, name=None, mutable=False,
         no_default_updates=False, accept_inplace=False, name=None,
         rebuild_strict=True, allow_input_downcast=None,
-        profile=None, on_unused_input=None):
+        profile=None, on_unused_input=None,output_keys=None):
```
pep8

lamblin commented on theano/compile/pfunc.py:
```diff
@@ -506,9 +506,9 @@ def pfunc(params, outputs=None, mode=None, updates=None, givens=None,
         inputs.append(si)
 
-    return orig_function(inputs, cloned_outputs, mode,
+    return orig_function(inputs, cloned_outputs, mode, 
```
pep8

DeathMonster666 commented:
Fixed issues related to str and basestring.  Added documentation for output_keys.  I'm not sure about the pep8 convention, since the existing code seems to violate it in a number of places.

lamblin commented:
About pep8, we started to enforce compliance on new files, but most existing files are in a whitelist.

delallea commented:
There's a test failure:

````
======================================================================
ERROR: Tests that an exception is thrown if a non-string key is used in
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/pyenv/lib/python3.3/site-packages/theano/tests/test_dictionary_output.py", line 154, in test_key_string_requirement
    theano.function([x], outputs={1.0: x, "a": x**2})
  File "/home/travis/miniconda/envs/pyenv/lib/python3.3/site-packages/theano/compile/function.py", line 192, in function
    output_items_sorted = sorted(output_items)
TypeError: unorderable types: str() < float()
````

DeathMonster666 commented:
That's a python 3 issue.  I can fix it soon.

nouiz commented:
There is this error left: 

```======================================================================

ERROR: Tests that an exception is thrown if a non-string key is used in

----------------------------------------------------------------------

Traceback (most recent call last):

File "/home/travis/miniconda/envs/pyenv/lib/python3.3/site-packages/theano/tests/test_dictionary_output.py", line 160, in test_key_string_requirement

theano.function([x], outputs={("a", "b"): x, 1.0: x**2})

File "/home/travis/miniconda/envs/pyenv/lib/python3.3/site-packages/theano/compile/function.py", line 192, in function

output_items_sorted = sorted(output_items)

TypeError: unorderable types: tuple() < float()

----------------------------------------------------------------------

Ran 1407 tests in 1579.714s
```

I think after that, it is good to merge.

DeathMonster666 commented:
Fixed issue.  Passes test now.

delallea commented:
I have no further comment, thanks!

nouiz commented on theano/compile/function.py:
```diff
@@ -185,6 +186,22 @@ def opt_log1p(node):
+
+        output_keys = []
+        outputs = []
+        for pair in output_items_sorted: 
```
Why python 3 raise a different error is because you make the assert after the call to `sorted`. If you make the check for string before that call, all python version will raise an assertion error. You keep the creation of output_keys here.

Can you do this change to have the same behavior with python 2 and 3?

nouiz commented:
Thanks for pushing this. Normally, there isn't so many details. You just seem to hit a spot that is more sensitive. You can see how much interrest in this PR, with the number of reviewer this PR have :) Normally, it isn't as much.

lamblin commented:
I think all my comments have been addressed, and it looks good to me.
If noone else has comments, I'll try to merge tomorrow.

nouiz commented:
I'm doing the NEWS.txt since the last release. I found that this functionality isn't documented in the tutorial of easy to find place. If someone want to add that, it could be great.
